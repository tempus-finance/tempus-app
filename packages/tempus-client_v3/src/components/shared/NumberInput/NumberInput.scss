.tc__text-input {
  .tc__text-input__container {
    input {
      border-radius: 4px 0 0 4px;
    }
    .tc__btn {
      width: 54px;
      height: 38px;
      border-radius: 0 4px 4px 0;
      background: var(--textInputBackground);
      border-left: 1px solid var(--textInputBorder);
    }
  }

  &.tc__text-input__disabled {
    .tc__text-input__container {
      .tc__btn {
        background: var(--textInputBackgroundDisabled);
        border-left: 1px solid var(--numberInputMidBorderDisabled);
      }
    }
  }
}
