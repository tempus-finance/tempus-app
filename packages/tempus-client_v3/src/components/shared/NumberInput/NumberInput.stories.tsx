import { ComponentStory } from '@storybook/react';
import React, { FC, useCallback, useState } from 'react';
import NumberInput, { NumberInputProps } from './NumberInput';

export default {
  title: 'NumberInput',
  component: NumberInput,
  argTypes: {},
};

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

const Wrapper: FC<NumberInputProps> = props => {
  const [value, setValue] = useState<string>(props.value ?? '');
  const onChange = useCallback((val: string) => setValue(val), []);
  return <NumberInput {...props} value={value} onChange={onChange} />;
};

const Template: ComponentStory<typeof NumberInput> = args => (
  <div style={style}>
    <Wrapper {...args} />
  </div>
);

export const NumberInputRaw = Template.bind({});
NumberInputRaw.args = {
  placeholder: 'placeholder',
  max: 100,
  precision: 0,
  disabled: false,
  debounce: false,
};

export const NumberInputNormal = Template.bind({});
NumberInputNormal.args = {
  label: 'number input',
  placeholder: 'placeholder',
  max: 100,
  precision: 0,
  disabled: false,
  debounce: false,
};

export const NumberInputCaption = Template.bind({});
NumberInputCaption.args = {
  label: 'number input',
  placeholder: 'placeholder',
  caption: 'caption here',
  max: 100,
  precision: 0,
  disabled: false,
  debounce: false,
};

export const NumberInputError = Template.bind({});
NumberInputError.args = {
  label: 'number input',
  placeholder: 'placeholder',
  caption: 'caption here',
  error: 'error here',
  max: 100,
  precision: 0,
  disabled: false,
  debounce: false,
};

export const NumberInputDisabled = Template.bind({});
NumberInputDisabled.args = {
  label: 'number input',
  placeholder: 'placeholder',
  caption: 'caption here',
  max: 100,
  precision: 0,
  disabled: true,
  debounce: false,
};
