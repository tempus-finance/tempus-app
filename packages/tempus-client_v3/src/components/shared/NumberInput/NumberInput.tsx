import React, { FC, memo, useCallback, useMemo } from 'react';
import TextInput from '../TextInput';
import ButtonWrapper from '../ButtonWrapper';
import Typography from '../Typography';

import './NumberInput.scss';

export interface NumberInputProps {
  label?: string;
  value?: string;
  max?: number;
  placeholder?: string;
  caption?: string;
  error?: string;
  disabled?: boolean;
  debounce?: boolean | number;
  onChange?: (value: string) => void;
}

const NumberInput: FC<NumberInputProps> = props => {
  const { label, value, max, placeholder, caption, error, disabled, debounce, onChange } = props;

  const onMaxClick = useCallback(() => onChange?.(`${max}`), [max, onChange]);
  const maxButton = useMemo(
    () => (
      <ButtonWrapper disabled={disabled} onClick={onMaxClick}>
        <Typography variant="body-primary" weight="regular" color={disabled ? 'text-disabled' : undefined}>
          Max
        </Typography>
      </ButtonWrapper>
    ),
    [max, disabled, onMaxClick],
  );

  return (
    <TextInput
      label={label}
      value={value}
      placeholder={placeholder}
      pattern="[0-9,]*[.]?[0-9]*"
      caption={caption}
      error={error}
      disabled={disabled}
      debounce={debounce}
      endAdornment={maxButton}
      onChange={onChange}
    />
  );
};

export default memo(NumberInput);
