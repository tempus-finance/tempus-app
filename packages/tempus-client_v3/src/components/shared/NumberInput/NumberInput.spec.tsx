import { fireEvent, render } from '@testing-library/react';
import NumberInput, { NumberInputProps } from './NumberInput';

const mockOnChange = jest.fn();

const defaultProps: NumberInputProps = {
  label: 'number input label',
  value: '',
  max: 100,
  placeholder: 'number input placeholder',
  caption: '',
  error: '',
  disabled: false,
  onChange: mockOnChange,
};

const subject = (props: NumberInputProps) => render(<NumberInput {...props} />);

describe('NumberInput', () => {
  it('renders a number input with label', () => {
    const { container, getByRole, queryByLabelText } = subject(defaultProps);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    expect(numberInput).toMatchSnapshot();
  });

  it('renders a number input with value', () => {
    const props = { ...defaultProps, value: 'value' };
    const { container, getByRole, queryByLabelText } = subject(props);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();
    expect(input).toHaveValue(props.value);

    expect(numberInput).toMatchSnapshot();
  });

  it('renders a number input with pattern', () => {
    const props = { ...defaultProps, pattern: '[0-9,]*[.]?[0-9]*' };
    const { container, getByRole, queryByLabelText } = subject(props);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();
    expect(input.getAttribute('pattern')).toBe(props.pattern);

    expect(numberInput).toMatchSnapshot();
  });

  it('renders a number input with caption', () => {
    const props = { ...defaultProps, caption: 'this is a caption' };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);
    const caption = queryByText(props.caption);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();
    expect(caption).not.toBeNull();

    expect(numberInput).toMatchSnapshot();
  });

  it('renders a number input with caption and error', () => {
    const props = { ...defaultProps, caption: 'this is a caption', error: 'this is an error' };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);
    const caption = queryByText(props.caption);
    const error = queryByText(props.error);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();
    expect(caption).toBeNull();
    expect(error).not.toBeNull();

    expect(numberInput).toMatchSnapshot();
  });

  it('renders a disabled number input', () => {
    const props = { ...defaultProps, disabled: true };
    const { container, getByRole, queryByLabelText } = subject(props);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();
    expect(input.getAttribute('disabled')).not.toBeNull();

    expect(numberInput).toMatchSnapshot();
  });

  it('renders a number input with focus', () => {
    const { container, getByRole, queryByLabelText } = subject(defaultProps);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.focus(input);

    expect(numberInput).toMatchSnapshot();
  });

  it('renders a number input with focus with caption', () => {
    const props = { ...defaultProps, caption: 'this is a caption' };
    const { container, getByRole, queryByText, queryByLabelText } = subject(props);

    const numberInput = container.querySelector('.tc__text-input');
    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);
    const caption = queryByText(props.caption);

    expect(numberInput).not.toBeNull();
    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();
    expect(caption).not.toBeNull();

    fireEvent.focus(input);

    expect(numberInput).toMatchSnapshot();
  });

  it('type text in number input will trigger onChange', () => {
    const { getByRole, queryByLabelText } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: '1234' } });

    expect(mockOnChange).toHaveBeenCalledWith('1234');
  });

  it('click max button in number input will trigger onChange', () => {
    const { getByRole, queryByLabelText } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.click(button);

    expect(mockOnChange).toHaveBeenCalledWith(`${defaultProps.max}`);
  });

  it('type text with unmatch pattern in number input will not trigger onChange', () => {
    const { getByRole, queryByLabelText } = subject(defaultProps);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });

    expect(mockOnChange).not.toHaveBeenCalled();
  });

  it('type text in number input with zero debounce will trigger onChange immediately', () => {
    const props = { ...defaultProps, debounce: 0 };
    const { getByRole, queryByLabelText } = subject(props);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: '1234' } });

    expect(mockOnChange).toHaveBeenCalledWith('1234');
  });

  it('type text in number input with debounce will trigger onChange after 300ms', () => {
    jest.useFakeTimers();
    const props = { ...defaultProps, debounce: true };
    const { getByRole, queryByLabelText } = subject(props);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: '1234' } });

    expect(mockOnChange).not.toHaveBeenCalled();
    jest.advanceTimersByTime(300);
    expect(mockOnChange).toHaveBeenCalledWith('1234');
    jest.useRealTimers();
  });

  it('type text in number input with debounce 1000 will trigger onChange after 1000ms', () => {
    jest.useFakeTimers();
    const props = { ...defaultProps, debounce: 1000 };
    const { getByRole, queryByLabelText } = subject(props);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: '1234' } });

    expect(mockOnChange).not.toHaveBeenCalled();
    jest.advanceTimersByTime(300);
    expect(mockOnChange).not.toHaveBeenCalled();
    jest.advanceTimersByTime(700);
    expect(mockOnChange).toHaveBeenCalledWith('1234');
    jest.useRealTimers();
  });

  it('keep typing text in number input with debounce will trigger onChange after 300ms from stop typing', () => {
    jest.useFakeTimers();
    const props = { ...defaultProps, debounce: true };
    const { getByRole, queryByLabelText } = subject(props);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: '1' } });

    expect(mockOnChange).not.toHaveBeenCalled();
    jest.advanceTimersByTime(200);

    fireEvent.change(input, { target: { value: '12' } });

    expect(mockOnChange).not.toHaveBeenCalled();
    jest.advanceTimersByTime(200);

    fireEvent.change(input, { target: { value: '123' } });

    expect(mockOnChange).not.toHaveBeenCalled();
    jest.advanceTimersByTime(200);

    fireEvent.change(input, { target: { value: '1234' } });

    expect(mockOnChange).not.toHaveBeenCalled();
    jest.advanceTimersByTime(300);

    expect(mockOnChange).toHaveBeenCalledWith('1234');
    jest.useRealTimers();
  });

  it('blur in number input with debounce will trigger onChange immediately', () => {
    jest.useFakeTimers();
    const props = { ...defaultProps, value: '1234', debounce: true };
    const { getByRole, queryByLabelText } = subject(props);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.blur(input);

    expect(mockOnChange).toHaveBeenCalledWith('1234');
    jest.useRealTimers();
  });

  it('type text and then blur in number input with debounce will trigger onChange immediately', () => {
    jest.useFakeTimers();
    const props = { ...defaultProps, value: undefined, debounce: true };
    const { getByRole, queryByLabelText } = subject(props);

    const input = getByRole('textbox');
    const button = getByRole('button');
    const label = queryByLabelText(defaultProps.label as string);

    expect(input).not.toBeNull();
    expect(button).not.toBeNull();
    expect(label).not.toBeNull();

    fireEvent.change(input, { target: { value: '1234' } });
    jest.advanceTimersByTime(100);
    fireEvent.blur(input);

    expect(mockOnChange).toHaveBeenCalledWith('1234');
    jest.useRealTimers();
  });
});
