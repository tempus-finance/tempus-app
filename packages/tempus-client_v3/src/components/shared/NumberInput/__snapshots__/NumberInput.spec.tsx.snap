// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput renders a disabled number input 1`] = `
<div
  class="tc__number-input"
>
  <div
    class="tc__text-input tc__text-input__disabled"
  >
    <label
      for="text-input-2"
    >
      <div
        style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        number input label
      </div>
    </label>
    <div
      class="tc__text-input__container"
    >
      <div
        style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        <input
          class="tc__base-input"
          disabled=""
          id="text-input-2"
          pattern="[0-9,]*[.]?[0-9]*"
          placeholder="number input placeholder"
          type="text"
          value=""
        />
      </div>
      <button
        class="tc__btn "
        data-selected="false"
        disabled=""
        type="button"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Max
        </div>
      </button>
    </div>
    
    
  </div>
</div>
`;

exports[`NumberInput renders a number input with label, value and caption 1`] = `
<div
  class="tc__number-input"
>
  <div
    class="tc__text-input "
  >
    <label
      for="text-input-1"
    >
      <div
        style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        number input label
      </div>
    </label>
    <div
      class="tc__text-input__container"
    >
      <div
        style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        <input
          class="tc__base-input"
          id="text-input-1"
          pattern="[0-9,]*[.]?[0-9]*"
          placeholder="number input placeholder"
          type="text"
          value="value"
        />
      </div>
      <button
        class="tc__btn "
        data-selected="false"
        type="button"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Max
        </div>
      </button>
    </div>
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      this is a caption
    </div>
    
  </div>
</div>
`;
